lib/Pipeline/Runner.cpp

```diff
@@ -291,32 +293,33 @@ llvm::Expected<DiffMap> Runner::runAllAnalyses() {
 Error Runner::run(llvm::StringRef EndingStepName,
                   const ContainerToTargetsMap &Targets) {
 
-  ContainerToTargetsMap ToLoad;
   vector<PipelineExecutionEntry> ToExec;
 
-  if (auto
-        Error = getObjectives(*this, EndingStepName, Targets, ToLoad, ToExec);
-      Error)
+  if (auto Error = getObjectives(*this, EndingStepName, Targets, ToExec); Error)
     return Error;
 
-  explainPipeline(Targets, ToLoad, ToExec);
+  explainPipeline(Targets, ToExec);
 
   if (ToExec.size() <= 1)
     return Error::success();
 
-  auto &FirstStepContainers = ToExec.front().ToExecute->containers();
-  auto CurrentContainer(FirstStepContainers.cloneFiltered(ToLoad));
-  for (auto &StepGoalsPairs :
-       llvm::make_range(ToExec.begin() + 1, ToExec.end())) {
-    auto &[Step, Goals] = StepGoalsPairs;
-    CurrentContainer = Step->cloneAndRun(*TheContext,
-                                         std::move(CurrentContainer));
+  for (auto &StepGoalsPairs : llvm::drop_begin(ToExec)) {
+    auto &[Step, PredictedOutput, Input] = StepGoalsPairs;
+    auto &Parent = Step->getPredecessor();
+    auto CurrentContainer = Parent.containers().cloneFiltered(Input);
+    auto Produced = Step->cloneAndRun(*TheContext, std::move(CurrentContainer));
+    revng_check(Produced.enumerate().contains(PredictedOutput),
+                "predicted output was not fully contained in actually "
+                "produced");
+    revng_check(Step->containers().enumerate().contains(PredictedOutput));
   }
 
   ExplanationLogger << "PRODUCED \n";
   indent(ExplanationLogger, 1);
   ExplanationLogger << EndingStepName << ":\n";
-  CurrentContainer.enumerate().dump(ExplanationLogger, 2, false);
+  ToExec.back().ToExecute->containers().enumerate().dump(ExplanationLogger,
+                                                         2,
+                                                         false);
   ExplanationLogger << DoLog;
```

Comment:
Wrap in `if (ExplanationLogger.isEnabled())`. Clearly isolating logging logic makes it easier to read the code.